### PR TITLE
Implement AutoEvent mechanism

### LINKED ISSUE #215: Implement AutoEvent mechanism
Device Services should be able to polling resources according to AutoEvent of Device model

### PR DESCRIPTION
**Pull request by cloudxxx8**
Create autoevent package
Trigger AutoEvent when starting up
Restart AutoEvents when the device is updated
Start AutoEvents when new device is created
Stop AutoEvents when the device is deleted or service is stopped
Add AutoEvent sample to device-simple
fix #215

Signed-off-by: Cloud Tsai <cloudxxx8@gmail.com>

### CODE REVIEW

**Comment by cloudxxx8** [comment]
recheck

File: internal/autoevent/executor.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import (
+	"fmt"
+	"time"
+
+	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/internal/handler"
+	"github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+const millionSecPerDay uint32 = 86400000
```
Comment (tonyespy) [inline internal/autoevent/executor.go]: Suggest "milliSecPerDay" or "msPerDay"

File: internal/autoevent/executor.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+type Executor interface {
+	Run()
+	Stop()
+}
+
+type executor struct {
+	deviceName   string
+	autoEvent    models.AutoEvent
+	lastReadings map[string]string
```
Comment (tonyespy) [inline internal/autoevent/executor.go]: Note, you'll also need a separate map for binary readings.

File: internal/autoevent/executor.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		}
+		time.Sleep(e.duration)
+
+		common.LoggingClient.Debug(fmt.Sprintf("AutoEvent - executing %v", e.autoEvent))
+		evt, appErr := readResource(e)
+		if appErr != nil {
+			common.LoggingClient.Error(fmt.Sprintf("AutoEvent - error occurs when read resource %s",
+				e.autoEvent.Resource))
+			continue
+		} else if evt != nil {
```
Comment (tonyespy) [inline internal/autoevent/executor.go]: Is there a scenario where evt and appErr can both be nil?

File: internal/autoevent/executor.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+				}
+				cacheReadings(e, evt.Readings)
+			}
+			common.LoggingClient.Debug(fmt.Sprintf("AutoEvent - pushing event %s", evt.String()))
+			go common.SendEvent(evt)
+		}
+	}
+}
+
+func readResource(e *executor) (*models.Event, common.AppError) {
```
Comment (tonyespy) [inline internal/autoevent/executor.go]: Why isn't this, and rest of functions that just take **e** as a parameter, methods?

File: internal/autoevent/executor.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func cacheReadings(e *executor, readings []models.Reading) {
+	for _, r := range readings {
+		e.lastReadings[r.Name] = r.Value
+	}
+}
+
+func compareReadings(e *executor, readings []models.Reading) bool {
+	for _, r := range readings {
```
Comment (tonyespy) [inline internal/autoevent/executor.go]: The way the caching works here is subtly different than that offered by the Java code which cached on a per deviceResource basis. Here, if any of the readings differ from last time around, then the entire event is sent. This probably makes sense, however at minimum this should be documented in a comment.

File: internal/autoevent/executor.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	return true
+}
+
+func (e *executor) Stop() {
+	e.stop = true
+}
+
+func NewExecutor(deviceName string, ae models.AutoEvent) Executor {
+	// check Frequency
+	duration := (time.Duration)(millionSecPerDay/ae.Frequency) * time.Millisecond
```
Comment (tonyespy) [inline internal/autoevent/executor.go]: From the code, it looks like Frequency has to specified in milliseconds, however the godoc in core-contracts doesn't say this (it should). Why aren't we using the standard time package duration formats (e.g. 1h, 2m, ...)? What does Support Scheduler support (some form of ITEF or ISO standard for specifying times)?

File: internal/autoevent/executor.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+		}
+	}
+	return true
+}
+
+func (e *executor) Stop() {
+	e.stop = true
+}
+
+func NewExecutor(deviceName string, ae models.AutoEvent) Executor {
```
Comment (tonyespy) [inline internal/autoevent/executor.go]: Shouldn't internal code still have godoc for public functions?

File: internal/cache/profiles.go [hunk 8]
```diff
@@ -174,11 +171,11 @@ func (p *profileCache) ResourceOperations(profileName string, cmd string, method
 	var resOps []models.ResourceOperation
 	var rosMap map[string][]models.ResourceOperation
 	var ok bool
-	if strings.ToLower(method) == getOpsStr {
+	if strings.ToLower(method) == common.GetCmdMethod {
```
Comment (tonyespy) [inline internal/cache/profiles.go]: This is one of changes unrelated to AutoEvents that I mention in my global review comments. At minimum this should be in it's own commit, but really a separate PR is more appropriate here.

File: internal/cache/profiles_test.go [hunk 9]
```diff
@@ -158,51 +158,51 @@ func TestProfileCache_CommandExists(t *testing.T) {
 func TestProfileCache_ResourceOperations(t *testing.T) {
 	dpc := newProfileCache(dps)
 
-	if _, err := dpc.ResourceOperations(mock.NewDeviceProfile.Name, mock.NewDeviceProfile.Commands[0].Name, getOpsStr); err == nil {
+	if _, err := dpc.ResourceOperations(mock.NewDeviceProfile.Name, mock.NewDeviceProfile.Commands[0].Name, common.GetCmdMethod); err == nil {
```
Comment (tonyespy) [inline internal/cache/profiles_test.go]: Same comment as for profiles.go applies to this file.

File: internal/cache/valuedescriptors.go [hunk 10]
```diff
@@ -9,6 +9,7 @@ package cache
 
 import (
 	"fmt"
+
```
Comment (tonyespy) [inline internal/cache/valuedescriptors.go]: Again, this is just whitespace cleanup... Either put all of your whitespace changes into a single commit, or as with the const changes, a separate PR is more appropriate.

File: internal/common/utils.go [hunk 11]
```diff
@@ -211,38 +209,6 @@ func CompareStrStrMap(a map[string]string, b map[string]string) bool {
 	return true
 }
 
-func MakeAddressable(name string, addr *models.Addressable) (*models.Addressable, error) {
```
Comment (tonyespy) [inline internal/common/utils.go]: Same as const, and whitespace changes. At minimum this belongs in its own commit ("rm unused Addressable code"), or separate PR.

File: internal/controller/restfuncs.go [hunk 12]
```diff
@@ -76,7 +78,7 @@ func callbackFunc(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	appErr := handler.CallbackHandler(cbAlert, req.Method)
+	appErr := callback.CallbackHandler(cbAlert, req.Method)
```
Comment (tonyespy) [inline internal/controller/restfuncs.go]: Why did callback need its own package?

File: internal/handler/command.go [hunk 13]
```diff
@@ -25,15 +25,15 @@ import (
 // Note, every HTTP request to ServeHTTP is made in a separate goroutine, which
 // means care needs to be taken with respect to shared data accessed through *Server.
 func CommandHandler(vars map[string]string, body string, method string) (*models.Event, common.AppError) {
-	dKey := vars["id"]
-	cmd := vars["command"]
+	dKey := vars[common.IdVar]
```
Comment (tonyespy) [inline internal/handler/command.go]: See above comments about const re-naming...

File: internal/handler/command.go [hunk 14]
```diff
@@ -176,10 +176,9 @@ func execReadCmd(device *models.Device, cmd string) (*models.Event, common.AppEr
 		return nil, common.NewServerError(msg, nil)
 	}
 
-	// push to Core Data
+	// generate event
 	event := &models.Event{Device: device.Name, Readings: readings}
 	event.Origin = time.Now().UnixNano() / int64(time.Millisecond)
-	go common.SendEvent(event)
```
Comment (tonyespy) [inline internal/handler/command.go]: This should be in a different commit than all the const changes!

File: internal/controller/restfuncs.go [hunk 15]
```diff
@@ -100,6 +102,9 @@ func commandFunc(w http.ResponseWriter, req *http.Request) {
 	if appErr != nil {
 		http.Error(w, fmt.Sprintf("%s %s", appErr.Message(), req.URL.Path), appErr.Code())
 	} else if event != nil {
+		// push to Core Data
+		go common.SendEvent(event)
```
Comment (tonyespy) [inline internal/controller/restfuncs.go]: Why did this get moved from command.go? Where we only sending events to Core Data for reads?

File: internal/handler/command_test.go [hunk 16]
```diff
@@ -18,7 +18,7 @@ func TestParseWriteParamsWrongParamName(t *testing.T) {
 
 	_, err := parseWriteParams(profileName, roMap, params)
 
-	if  err == nil {
+	if err == nil {
```
Comment (tonyespy) [inline internal/handler/command_test.go]: whitespace only changes in this file (see above)!

File: internal/mock/mock_deviceclient.go [hunk 17]
```diff
@@ -16,7 +16,7 @@ import (
 )
 
 const (
-	InvalidDeviceId   = "1ef435eb-5060-49b0-8d55-8d4e43239800"
+	InvalidDeviceId = "1ef435eb-5060-49b0-8d55-8d4e43239800"
```
Comment (tonyespy) [inline internal/mock/mock_deviceclient.go]: whitespace only changes in this file (see above)!

File: internal/mock/mock_deviceprofileclient.go [hunk 18]
```diff
@@ -19,7 +19,7 @@ var (
 	DeviceProfileRandomUnsignedGenerator       = models.DeviceProfile{}
 	DeviceProfileRandomFloatGenerator          = models.DeviceProfile{}
 	DuplicateDeviceProfileRandomFloatGenerator = models.DeviceProfile{}
-	NewDeviceProfile 							= models.DeviceProfile{}
+	NewDeviceProfile                           = models.DeviceProfile{}
```
Comment (tonyespy) [inline internal/mock/mock_deviceprofileclient.go]: whitespace only changes in this file (see above)!

File: internal/cache/profiles.go (hunk 8 above)
Comment (cloudxxx8) [reply internal/cache/profiles.go]: will do

File: internal/cache/profiles_test.go (hunk 9 above)
Comment (cloudxxx8) [reply internal/cache/profiles_test.go]: will do

File: internal/cache/valuedescriptors.go (hunk 10 above)
Comment (cloudxxx8) [reply internal/cache/valuedescriptors.go]: will do

File: internal/common/utils.go (hunk 11 above)
Comment (cloudxxx8) [reply internal/common/utils.go]: will do

File: internal/controller/restfuncs.go (hunk 12 above)
Comment (cloudxxx8) [reply internal/controller/restfuncs.go]: we need it to resolve the recursive dependency issue
autoevent uses handler, and callback needs to update autoevent

File: internal/handler/command.go (hunk 13 above)
Comment (cloudxxx8) [reply internal/handler/command.go]: will do

File: internal/handler/command.go (hunk 14 above)
Comment (cloudxxx8) [reply internal/handler/command.go]: move this line for the AutoEvent OnChange mechanism, or the event will be sent here anyway

File: internal/controller/restfuncs.go (hunk 15 above)
Comment (cloudxxx8) [reply internal/controller/restfuncs.go]: move this line for the AutoEvent OnChange mechanism, or the event will be sent in command handler anyway

File: internal/handler/command_test.go (hunk 16 above)
Comment (cloudxxx8) [reply internal/handler/command_test.go]: white space changed by go fmt

File: internal/mock/mock_deviceclient.go (hunk 17 above)
Comment (cloudxxx8) [reply internal/mock/mock_deviceclient.go]: white space changed by go fmt

File: internal/mock/mock_deviceprofileclient.go (hunk 18 above)
Comment (cloudxxx8) [reply internal/mock/mock_deviceprofileclient.go]: white space changed by go fmt

File: example/cmd/device-simple/res/configuration.toml [hunk 19]
```diff
@@ -65,3 +65,7 @@ File = "./device-simple.log"
     [DeviceList.Protocols.other]
       Address = "simple01"
       Port = "300"
+  [[DeviceList.AutoEvents]]
+    Frequency = 8640
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply example/cmd/device-simple/res/configuration.toml]: ok, will do

File: internal/autoevent/executor.go (hunk 1 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: ok, will do

File: internal/autoevent/executor.go (hunk 2 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: we could create another PR after CBOR implementation completed

File: internal/autoevent/executor.go (hunk 4 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: I will refactor it to make them consistent

File: internal/autoevent/executor.go (hunk 3 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: No, I don't think so.  Should we add this check?

File: internal/autoevent/executor.go (hunk 5 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: I would limit it to only send the changed readings.  We could discuss in WG meeting today.

File: internal/autoevent/executor.go (hunk 6 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: Using standard time package might not compatible with C SDK.
We could discuss the Frequency format in WG meeting

Comment (cloudxxx8) [reply internal/autoevent/executor.go]: it is not necessary to use million second.  I just always calculate it to million second.

File: internal/autoevent/executor.go (hunk 7 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: ok, will add

**Comment by cloudxxx8** [comment]
@tonyespy , thanks for the review comments, and I will modify it.
I thought we will have some topics to discuss in WG meeting.

For the cache you mentioned, since OnChange field is under AutoEvent only, there is no way to take effect on REST calls to the /command endpoint.

File: example/cmd/device-simple/res/configuration.toml (hunk 19 above)
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply example/cmd/device-simple/res/configuration.toml]: time.ParseDuration(durCountdown)

**Comment by cloudxxx8** [comment]
Opened PR for Frequency data model update
https://github.com/edgexfoundry/go-mod-core-contracts/pull/50

File: internal/autoevent/executor.go (hunk 5 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: According to WG meeting, we would keep it as is

**Comment by iain-anderson** [comment]
See also https://github.com/edgexfoundry/go-mod-core-contracts/issues/51

**Comment by akramtexas** [comment]
- A detailed discussion regarding the implementation of the AutoEvent mechanism took place during the Device Services Working Group meeting (on 25-Mar-2019). Captured thoughts in the meeting minutes on Wiki.

File: internal/autoevent/executor.go [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+		common.LoggingClient.Debug(fmt.Sprintf("AutoEvent - executing %v", e.autoEvent))
+		evt, appErr := readResource(e)
+		if appErr != nil {
+			common.LoggingClient.Error(fmt.Sprintf("AutoEvent - error occurs when read resource %s",
+				e.autoEvent.Resource))
+			continue
+		} else if evt != nil {
+			if e.autoEvent.OnChange {
+				if compareReadings(e, evt.Readings) {
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: @tobiasmo1 sicne OnChange filed only in AutoEvent, we would need find another way to implement the further OnChange mechanism for normal Get command.
Annotating the event with a 'changed' field is another topic, and we should discuss in WG meeting.

**Comment by cloudxxx8** [comment]
@tonyespy @iain-anderson 
I've modified this PR and pushed
Please help review again
For the new callback package, I will open another PR to refactor it by separating the code to different go file, because the current callback.go is too large and it causes the recursive package issue in this PR

File: internal/autoevent/executor.go (hunk 3 above)
Comment (tonyespy) [reply internal/autoevent/executor.go]: For starters, the go style isn't to chain if/else code like this. Instead it should be:
```

if appErr != nil {
}

if evt != nil {
}

if evt == nil {
}

// handle evt
```

File: internal/controller/restfuncs.go (hunk 15 above)
Comment (tonyespy) [reply internal/controller/restfuncs.go]: This is my biggest concern.  This actually now means we're encoding event into JSON three times, unless I missed something...

If someone calls the command REST endpoint, then we encode the JSON once when `execReadCommand` creates the event. Then, we encode it in `commandFunc` with your new change which calls `common.SendEvent()`, and then again two lines later with: `json.NewEncoder(w).Encode(event)`

See issue #233

File: example/cmd/device-simple/res/configuration.toml [hunk 21]
```diff
@@ -65,3 +65,7 @@ File = "./device-simple.log"
     [DeviceList.Protocols.other]
       Address = "simple01"
       Port = "300"
+  [[DeviceList.AutoEvents]]
```
Comment (tonyespy) [inline example/cmd/device-simple/res/configuration.toml]: It's not clear to me with the given syntax how someone would add more than one device?

Comment (cloudxxx8) [reply example/cmd/device-simple/res/configuration.toml]: the syntax is the standard TOML configuration, and it is not relevant to this AutoEvent PR.  Please open another issue and we could enhance the configuration example to create multiple devices.

for the multiple encoding, since we have issue #233, why don't we resolve it in another PR?

File: internal/controller/restfuncs.go (hunk 15 above)
Comment (cloudxxx8) [reply internal/controller/restfuncs.go]: actually, it won't be three times, and it is a separated issue

File: internal/autoevent/executor.go (hunk 3 above)
Comment (cloudxxx8) [reply internal/autoevent/executor.go]: modified

File: example/cmd/device-simple/res/configuration.toml (hunk 21 above)
Comment (tonyespy) [reply example/cmd/device-simple/res/configuration.toml]: I just wanted you to confirm whether not the syntax supported multiple devices.

Re: multiple encoding let's keep that conversation in the relevant thread, this was about configuration only.

File: internal/autoevent/executor.go (hunk 4 above)
Comment (tonyespy) [reply internal/autoevent/executor.go]: Thanks for changing, although I did notice that cacheReadings and compareReadings still remain functions. I'm not going to block for this, just wanted to see if that was intentional.

**Comment by tonyespy** [comment]
I see you force-pushed the branch last night, but there are no new commits and you didn't explain what you changed????

@tobiasmo1 Can you please review my comment about multiple encodings. It's my contention that this code introduces a third encoding of the event, but Cloud claims I wrong about that. If I'm wrong, then I think we're OK to merge this PR (after Cloud explains the changes he pushed last night).

Unfortunately I'm going to miss Monday's WG meeting.

File: internal/controller/restfuncs.go (hunk 15 above)
Comment (tonyespy) [reply internal/controller/restfuncs.go]: How many encodings happen is evt != nil in the following code fragment:

```
} else if event != nil {
		// push to Core Data
		go common.SendEvent(event)

		w.Header().Set(clients.ContentType, clients.ContentTypeJSON)
		json.NewEncoder(w).Encode(event)
}
```

**Comment by cloudxxx8** [comment]
@tonyespy , sorry for the missing conversation, I only did the change for your first comment about 

```
if appErr != nil {
}

if evt != nil {
}

if evt == nil {
}
```

**Comment by cloudxxx8** [comment]
@tonyespy sorry again for my wrong description.  `go common.SendEvent(event)` should be removed, but I missed this change during the re-commit the changes.
Now, I removed this line from command.go, so it won't be encoded three times.

**Comment by tonyespy** [comment]
I've approved, but @iain-anderson needs to approve, as his last review requested changes.

Also it appears the arm build failed, so you need to kick that off once you get approval from Iain.

**Comment by cloudxxx8** [comment]
recheck

**Comment by cloudxxx8** [comment]
it looks like golang:1.11-alpine image is not available at this moment.  recheck tomorrow

> 10:50:45 Step 1/19 : FROM golang:1.11-alpine AS builder
> 10:50:46 1.11-alpine: Pulling from library/golang
> 10:50:46 no matching manifest for unknown in the manifest list entries
> 10:50:46 Makefile:21: recipe for target 'docker' failed
> 10:50:46 make: *** [docker] Error 1
> 10:50:46 Build step 'Execute shell' marked build as failure

**Comment by cloudxxx8** [comment]
recheck

**Comment by tsconn23** [comment]
recheck

File: internal/autoevent/executor.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (tobiasmo1) [reply internal/autoevent/executor.go]: Agree let's discuss annotating event as separate topic.

**Comment by tsconn23** [comment]
recheck
